refactor: Implement JobConfig as a NestJS module

sbliven opened the pull request:
<!--
Follow semantic-release guidelines for the PR title, which is used in the changelog.

Title should follow the format `<type>(<scope>): <subject>`, where
- Type is one of: build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test|BREAKING CHANGE
- Scope (optional) describes the place of the change (eg a particular milestone) and is usually omitted
- subject should be a non-capitalized one-line description in present imperative tense and not ending with a period

See https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines for more details.
-->

## Description
<!-- Short description of the pull request -->
Refactor the JobConfig following #1154.

## Motivation
<!-- Background on use case, changes needed -->
The main motivation is to make it easier to implement JobActions that depend on NestJS modules. For instance, `EmailJobAction` now can call MailService, `ValidateJobAction` will be able to make database calls (#1421 ), etc.

Registering JobActions was also awkward with the previous implementation, and tests were required to call some setup code. NestJS is great for managing singletons, so all this setup and caching can be replaced with standard module imports.

## Details

Implementing an Action now requires 4 (short) files:
- `action.ts` contains a class implementing `JobAction`. It's constructor can take any arguments, but the existing actions take an `options` argument mirroring the expected yaml config. It does not need to be `@Injectable()`, since it is constructed by the factory.
- `action.interface.ts` can contain additional types, e.g. the definition of the expected `options` and a type guard for casting to the options.
- `action.factory.ts` implements `JobActionFactory`. The factory is provided by NestJS as a singleton, so it must be `@Injectable()`. This means that dependencies can be injected into the factory. It has a `create(options)` method, which constructs the action itself by combining the options from the yaml file with any dependencies injected by NestJS.

The lists of known factories are provided to Nest with the `CREATE_JOB_ACTION_FACTORIES` and `UPDATE_JOB_ACTION_FACTORIES` symbols. The top-level AppModule imports built-in actions from `DefaultJobActionFactories`. Future plugins can then use the NestJS dependency injection system to extend the lists if needed.

Parsing jobConfig.yaml is handled by the JobConfigService. NestJS injects the lists of factories (and the file path) during construction. When parsing reaches a new action, the correct factory is used to create an instance of the JobAction with the current list of options.

Code which needs the configuration for a particular job type (eg jobs.controller.ts) injects the `JobConfigService` and can then call `jobConfigService.get(jobType)`.

## Changes:
<!-- Please provide a list of the changes implemented by this PR -->

* Refactor Job Configuration system (#1154)

## Tests included

- [ ] Included for each change/fix?
- [x] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [x] swagger documentation updated (required for API changes)
- [ ] official documentation updated

### official documentation info
<!-- If you have updated the official documentation, please provide PR # and URL of the updated pages -->

sbliven commented:
Unit tests should now pass again. I still need to confirm that mocha tests pass.

There are some lint errors due to `any` being used. I will fix these.

The rabbitmq and urlactions still need to be migrated.

sbliven committed:
Complete refactor of job configuration as NestJS modules

- Developing new actions is documented in README.md for developers
- Move everything related to job config to src/config/job-config
- Major refactor of parsing from jobConfig.yaml to JobConfig objects
- Create module, interface, and factory for each JobAction
- Add handlebar-utils to help consistently applying templates to jobs
  and Dtos.
  - In test contexts the handlebars helpers may not be registered, so
    avoid using custom helpers or register them in the test
- Add options to LogJobAction. This is also good for templating tests
- Change MailService.sendMail to match nest's MailerService.sendMail
- Accept empty jobConfigurationFile as no jobs
- CaslModule now depends indirectly on the MailerModule. Thus all
  controllers need to either mock CaslAbilityFactory or add a mock
  MailerModule before importing CaslModule.
- Most tests disable jobConfig except for those that directly test it.
  These use test/config/jobconfig.yaml

sbliven commented:
I've squashed all previous commits to make it easier to review.

Junjiequan commented on src/common/mail.service.ts:
```diff
@@ -1,28 +1,25 @@
+/**
+ * Service for sending emails using nodemailer.
+ *
+ * Use this rather than MailerService directly to allow configuration in AppModule
+ */
 @Injectable()
 export class MailService {
   constructor(private readonly mailerService: MailerService) {}
 
-  async sendMail(
-    to: string,
-    cc: string,
-    subject: string,
-    mailText: string | null,
-    html: string | null = null,
-  ) {
+  async sendMail(options: ISendMailOptions): Promise<SentMessageInfo> {
```
I see that we have `nodemailer` package installed separately, and it is used only as a type (which is Any type) in this file. However, `@nestjs-modules/mailer` already has nodemailer as part of the dependency package. 
I'd suggest to remove `nodemailer` from the package.json if we have no intention to use this package separately and rely on dependency nodemailer from @nestjs-modules/mailer exclusively

sbliven replied:
Good suggestion. I think `reflect-metadata` is also used only indirectly, but I will save removing that for a different PR.

nitrosx commented on src/auth/auth.module.ts:
```diff
@@ -13,7 +13,7 @@ import { OidcConfig } from "src/config/configuration";
 import { accessGroupServiceFactory } from "./access-group-provider/access-group-service-factory";
 import { AccessGroupService } from "./access-group-provider/access-group.service";
-import { CaslAbilityFactory } from "src/casl/casl-ability.factory";
+import { CaslModule } from "src/casl/casl.module";
```
I am curious about renaming the module?

sbliven replied:
It's not renamed. Importing the module gives you the app-wide singleton instance. The previous pattern was instantiating a new CaslAbilityFactory outside of the dependency injection framework.

nitrosx commented on src/datasets/datasets.module.ts:
What are the differences between using casl module as a providers and as an import? What are the advantages of the new implementation?

sbliven replied:
See above.

My understanding is that it is an error to directly instantiate a provider (eg anything marked `@Injectable` or `@Controller`). Instead you need to import the module and get the singleton from that module.

nitrosx reviewed (commented):
Just a couple of questions.
Other than that, I think this PR is good to go.

despadam commented on src/config/job-config/jobconfig.interface.ts:
```diff
@@ -0,0 +1,85 @@
+ */
+export interface JobConfig {
+  jobType: string;
+  configVersion: string;
```
Why do we also need `configVersion` per job? Isn't this information only on the file level?

sbliven replied:
This is how it was before. We don't store the top-level object (`JobConfigListOptions` here) so the root `configVersion` has to get copied into each JobConfig section. I'm not sure what happened before if you specified it down a level too. Maybe we should remove it from `JobConfigOptions` entirely, rather than using the top-level field as a default that can still be overridden one level down?

There's potentially a use case where you only update one job type and so only bump one version, but that sounds complicated to track.

despadam replied:
You are right that this is how it was before, I got confused. I think the confusion came from the fact that we also store `configVersion` in `JobConfigListOptions` but never actually access it from there. From what I see, we extract it directly either from `JobConfig` or `JobConfigOptions`, depending on the case.

We have agreed that we will not cover the use case of updating individually the versions, exactly because it becomes too tricky. There is only a check in the controller that will print an error message when a job's version is not the same as the current running config version - I believe this is enough

sbliven replied:
I've made `JobConfigListOptions.configVersion` required and made sure it always gets passed through to the JobConfig level.

The ajv schema enforces that configVersion can only appear as a top-level field.

despadam commented on src/config/job-config/jobconfig.interface.ts:
```diff
@@ -0,0 +1,85 @@
+  create: JobOperation<CreateJobDto>;
+  statusUpdate: JobOperation<StatusUpdateJobDto>;
+}
+export interface JobConfigOptions {
```
In general I am struggling a bit to understand this file. What is the difference between `JobConfig` and `JobConfigOptions` and why do we need both? The same with `JobOperation` etc ...

sbliven replied:
`JobConfigOptions` represents the POD object parsed from yaml. This is then converted to the real `JobConfig` instance. They have similar structure except that the `JobConfig` has `JobAction` instances at its leaves, while `JobConfigOptions` has the configuration strings as leaves.

sbliven replied:
It's possible this could be done directly in the yaml parser as a custom deserializer, but this follows the patterns used in Nest's documentation for custom providers.

despadam commented on src/config/job-config/jobconfig.service.ts:
```diff
@@ -0,0 +1,132 @@
+    const yaml = fs.readFileSync(filePath, "utf8");
+    const jobConfigListOptions = load(yaml, { filename: filePath });
+
+    // TODO: Do we like the schema-based validation, or should it be pure typescript?
```
For me it's ok to leave it as is. The schema only checks for minimal information as everything else can be configured. In case we decide to expand the schema, we could then change the validation approach. But for now I think it is not a concern

sbliven replied:
I'm not sure validation is even necessary, since all the constructors throw errors if the yaml doesn't follow the expected structure. I expect ajv gives clearer error messages though.

despadam commented on src/config/job-config/actions/defaultjobactions.module.ts:
```diff
@@ -0,0 +1,90 @@
+        emailJobActionFactory,
+        validateJobActionFactory,
+        urlJobActionFactory,
+        rabbitMQJobActionFactory,
```
Maybe I am misunderstanding this or the name `defaultjobactions` is confusing me, but by doing this don't we by default enable all actions that are written here? Instead we should only enable those that are declared in the jobConfig file. Perhaps if this was split to `default` and `additional` actions it would be clearer? Where in `additional` someone could declare whether the use rabbitMQ or Kafka for example, instead of by default having rabbitMQ

sbliven replied:
When the yaml file gets parsed you need to have a mapping between `actionType` strings and the `JobActionFactory` instance that will parse that part of the config and create a `JobAction`. Would it be clearer if the file were called `defaultjobactionfactories.module.ts` like the module? Somehow the name should reflect that this is the stuff that gets created before the yaml file is parsed, not the stuff configured by the yaml.

If you wanted to add new actions (say using a plugin) you would provide new lists for `CREATE_JOB_ACTION_FACTORIES` and `UPDATE_JOB_ACTION_FACTORIES`. I guess this could be split into the built-in ones and `additional` factories loaded from plugins, but it's not clear to me what you would gain from that.

sbliven replied:
I've renamed this to `JobActionCreator`.

despadam commented on src/config/job-config/actions/defaultjobactions.module.ts:
```diff
@@ -0,0 +1,90 @@
+  ],
+  exports: [CREATE_JOB_ACTION_FACTORIES, UPDATE_JOB_ACTION_FACTORIES],
+})
+export class DefaultJobActionFactories {}
```
Is this a factory or a module?

sbliven replied:
It's a module in the NestJS sense, but its role is to provide the factories-that-create-JobAction-instances. Please suggest better names, this is starting not to even make sense to me.

sbliven committed:
Remove nodemailer dependency

It is only needed indirectly through @nestjs-modules/mailer

despadam reviewed (commented):
Overall looks good, thanks for all the great work @sbliven! I left a few comments, plus have a general question: 
Going through these changes I realized that the controllers import factories, which in my understanding perhaps is not a great approach, as controllers should import services. More specifically, I am referring to the `CaslAbilityFactory` that keeps getting imported and reused multiple times. 
I think the topic of refactoring Casl has come up a few times already and I believe it is something that we will have to look into. But in anyway it should be in a separate PR, so we can discuss this at a later time

sbliven commented:
Thanks for the review @despadam.

> controllers import factories, which in my understanding perhaps is not a great approach, as controllers should import services. More specifically, I am referring to the CaslAbilityFactory that keeps getting imported and reused multiple times.

I think this is a problem with inconsistent naming. NestJS doesn't give clear definitions of all of these terms. Here's my own (possibly hallucinated by chatGPT) definitions:
- *provider*: A provider is a class or a value that can be injected into other classes (via dependency injection) to encapsulate reusable logic or state. Classes decorated with `@Injectable()` are providers, as are objects with a `{provides:token}`
- *service*:  A specific type of provider primarily used to define reusable business logic or data access operations. Every service is a provider, but not all providers are services.
- *module*: a namespace for dependency injection, responsible for instantiating providers, injecting them into other providers as needed, and exporting the singletons to other modules.
- *factory*: a function that generates a value or an object to be used as a provider. It can dynamically determine the provider’s instance at runtime based on logic or configuration.

Many of our classes don't follow these terms
- `CaslAbilityFactory` is actually a provider, not a factory. It is an `@Injectable()` class intended as a singleton
- `accessGroupServiceFactory` is a provider that uses an (anonymous) factory for its instantiation
- `JobActionFactory` subclasses are providers, since they use `@Injectable()`! They do create `JobAction` instances, but this is done outside of nestjs.

I played around a bit with trying to make JobActionFactories actual nest factories (eg functions that return providers). All that would be needed to make JobAction implementations into providers would be to add `@Injectable()`, so this seems within reach. But after trying to understand how dependency injection would work with a factory my brain threw a 'RecursionError: Maximum Recursion Depth Exceeded' and so I just committed the version here. I would be open to either suggestions as to changing the name or refactoring to make it an actual factory.

Junjiequan commented:
That's a very clear explaniation of the nestJs module system. Thanks @sbliven 

To add a little extra information when creating a module:
exports: Makes services/providers available for injection in other modules when this module is imported.
controllers: Declares which controllers should handle HTTP requests for the routes managed by this module.

sbliven committed:
Rename JobActionFactory to JobActionCreator

This avoids confusion with the nestJS 'factory' term.

- rename classes and instances
- rename all action.factory.ts to action.service.ts
- update developer README
- rename DefaultJobActionFactories to CoreJobActionCreators
- make top-level configVersion mandatory (this was already enforced
  by the ajv schema, but the ts type was more permissive)

sbliven committed:
lint:fix

sbliven commented:
In ec245bb I switch the code to use `JobActionCreator` to avoid confusion with the NestJS term `factory`. I suggest in #1560 renaming some of the other "factories" too, but for now I will only rename the ones introduced in this PR.
